### DISCUSSION TITLE
Problem with Benchmark Creation after Upgrade to 0.3.1

### DISCUSSION THREAD

**Discussion by jannishag** [ContinualAI/avalanche#1264] [Q&A] [answered]
Hi all,
I am experiencing the following error when creating an nc_benchmark using avalanche-lib 0.3.1:
AttributeError: ‘AvalancheDataset’ object has no attribute ‘targets’

I did not experience this using avalanche-lib 0.2.1, so maybe there is an issue with the update?
I would appreciate any help on that!

Here is the traceback:

---> 68         scenario = nc_benchmark(
     69             Train_List, Test_List, n_experiences=1, one_dataset_per_exp=True, shuffle=False, seed=1234, task_labels=True
     70         )

[/usr/local/lib/python3.8/dist-packages/avalanche/benchmarks/generators/benchmark_generators.py](https://uujlphuqgb-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20221216-060044-RC01_495938583#) in nc_benchmark(train_dataset, test_dataset, n_experiences, task_labels, shuffle, seed, fixed_class_order, per_exp_classes, class_ids_from_zero_from_first_exp, class_ids_from_zero_in_each_exp, one_dataset_per_exp, train_transform, eval_transform, reproducibility_data)
    194             seq_test_dataset,
    195             mapping,
--> 196         ) = concat_datasets_sequentially(train_dataset, test_dataset)
    197 
    198         if one_dataset_per_exp:

[/usr/local/lib/python3.8/dist-packages/avalanche/benchmarks/utils/utils.py](https://uujlphuqgb-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20221216-060044-RC01_495938583#) in concat_datasets_sequentially(train_dataset_list, test_dataset_list)
    202 
    203     # Obtain the number of classes of each dataset
--> 204     classes_per_dataset = [
    205         _count_unique(
    206             train_dataset_list[dataset_idx].targets,

[/usr/local/lib/python3.8/dist-packages/avalanche/benchmarks/utils/utils.py](https://uujlphuqgb-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20221216-060044-RC01_495938583#) in <listcomp>(.0)
    204     classes_per_dataset = [
    205         _count_unique(
--> 206             train_dataset_list[dataset_idx].targets,
    207             test_dataset_list[dataset_idx].targets,
    208         )

AttributeError: 'AvalancheDataset' object has no attribute 'targets'

EDIT:
From the API Docs I read that the attribute 'targets' was removed from AvalancheDatasets in the new release. Thus, I guess the utils.py has to be updated, right?

**Comment by AntonioCarta** [comment, marked as answer]
Hi @jannishag, can you give more details on your data? Without knowing your dataset we can't say anything about the error.

>  From the API Docs I read that the attribute 'targets' was removed from AvalancheDatasets in the new release. Thus, I guess the utils.py has to be updated, right?

We didn't remove the attribute, we made it optional. You can use `avalanche.benchmarks.utils.make_classification_dataset` to easily instantiate datasets with class and task labels. This is equivalent to the old `AvalancheDataset`, while the new `AvalancheDataset` is a generic object which doesn't necessarily have `targets`.

**Reply by jannishag** [reply]
Ah, alright! Thanks Antonio for the hint, I got it now.

If anyone experiences the same issue:
I did create my AvalancheDataset directly from a TensorDataset (code like, it was _AvalancheDataset(TensorDataset)_). Using avalanche.benchmarks.utils.make_classification_dataset instead of casting solved the problem.
